From 314b1e73c83eda37d48637f0cbb7daae9894e1b2 Mon Sep 17 00:00:00 2001
From: Jochen Breuer <jbreuer@suse.de>
Date: Thu, 13 Feb 2020 11:43:01 +0100
Subject: [PATCH] Adds missing function body for
 win32_kill_process_tree

This seems to have vanished with a previouy merge conflict.
---
 tests/support/helpers.py | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/tests/support/helpers.py b/tests/support/helpers.py
index 047752c3df..2cbab7d4b8 100644
--- a/tests/support/helpers.py
+++ b/tests/support/helpers.py
@@ -1626,6 +1626,27 @@ class MirrorPostHandler(tornado.web.RequestHandler):
 
 def win32_kill_process_tree(pid, sig=signal.SIGTERM, include_parent=True,
         timeout=None, on_terminate=None):
+    '''
+    Kill a process tree (including grandchildren) with signal "sig" and return
+    a (gone, still_alive) tuple.  "on_terminate", if specified, is a callabck
+    function which is called as soon as a child terminates.
+    '''
+    if pid == os.getpid():
+        raise RuntimeError("I refuse to kill myself")
+    try:
+        parent = psutil.Process(pid)
+    except psutil.NoSuchProcess:
+        log.debug("PID not found alive: %d", pid)
+        return ([], [])
+    children = parent.children(recursive=True)
+    if include_parent:
+        children.append(parent)
+    for p in children:
+        p.send_signal(sig)
+    gone, alive = psutil.wait_procs(children, timeout=timeout,
+                                    callback=on_terminate)
+    return (gone, alive)
+
 
 def this_user():
     '''
-- 
2.16.4


